src/high-level/abstract-tensor.lisp

```diff
@@ -0,0 +1,261 @@
+
+If LAYOUT is specified then traverse TENSOR in the specified order (column major or row major).")
+  (:method ((function function) (tensor abstract-tensor))
+    (into! function (deep-copy-tensor tensor))))
+
+(defgeneric sum (tensor)
+  (:documentation "Get the sum of the elements of TENSOR")
+  (:method ((tensor abstract-tensor))
+    (let ((sum 0))
+      (foreach (lambda (x) (incf sum x)) tensor)
+      sum)))
+
+(defgeneric scale (tensor factor)
+  (:documentation "Scale TENSOR by FACTOR, returning a new tensor of the same type as TENSOR")
+  (:method ((tensor abstract-tensor) (factor number))
+    (map! (lambda (x) (* x factor)) (deep-copy-tensor tensor))))
+
+(defgeneric scale! (tensor factor)
+  (:documentation "Scale TENSOR by FACTOR, storing back into the tensor")
+  (:method ((tensor abstract-tensor) (factor number))
+    (map! (lambda (x) (* x factor)) tensor)))
+
+(defgeneric slice (tensor from to)
+  (:documentation "Slice a tensor from FROM to TO, returning a new tensor with the contained elements")
+  (:method ((tensor abstract-tensor) from to)
+    (declare (type sequence from to))
+    (policy-cond:with-expectations (> speed safety)
+        ((assertion (and (valid-index-p from (shape tensor))
+                         (cl:every #'< from (shape tensor))))
+         (assertion (and (cl:= (order tensor) (length to))
+                         (valid-shape-p to)
+                         (cl:every #'<= to (shape tensor))))
+         (assertion (cl:every #'<= from to)))
+      (let* ((dims (mapcar #'- to from))
+             (target (empty dims
+                            :layout (layout tensor)
+                            :type (element-type tensor))))
+        (map-indexes dims
+                     (lambda (&rest dims)
+                       (setf (apply #'tref target dims)
+                             (apply #'tref tensor (mapcar #'+ dims from)))))
+        target))))
+
+(defgeneric binary-operator (function source1 source2 &optional target)
+  (:documentation "Perform a binary operator on tensors elementwise, optionally storing the result in TARGET.
+If TARGET is not specified then a new tensor is created with the same element type as the first source tensor")
+  (:method ((function function) (source1 abstract-tensor) (source2 abstract-tensor) &optional target)
+    (policy-cond:with-expectations (> speed safety)
+        ((assertion (equalp (shape source1) (shape source2))))
+      (let ((target (or target (copy-tensor source1))))
+        (map-indexes
+         (shape source1)
+         (lambda (&rest dims)
+           (apply #'(setf tref)
+                  (funcall function
+                           (apply #'tref source1 dims)
+                           (apply #'tref source2 dims))
+                  target dims)))
+        target))))
+
+(defgeneric .+ (source1 source2 &optional target)
```

Comment:
For MAGICL v0.9.999 or v1.0.001 we might want to drop the dots in front of the operators. There is no confusion on what the meaning of `+` or `-` is when applied to tensors. Confusion only occurs with multiplication (choice among tensor-tensor or element-wise product), division (here the dichotomy occurs between multiplying by the inverse or doing element-wise division), and exponentiation.

Reply:
I agree on this one. Do you think this would be good to push through in this PR or hold off for later. My only hesitation on holding off until later is that quilc might need another PR and release cycle to clear it up (there are five uses of `.+` in quilc.)

Reply:
I'm OK either way but I'll defer to @stylewarning on this.

Reply:
I'll defer back to @jmbr

Reply:
Alright. I'm OK with @colescott going ahead with this in its current state.